subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/PropertyValue.java

```diff
@@ -17,15 +17,28 @@
 package org.gradle.api.internal.tasks.properties;
 
 import org.gradle.api.Task;
+import org.gradle.api.provider.Provider;
 
 import javax.annotation.Nullable;
 import java.util.concurrent.Callable;
 
 public interface PropertyValue extends Callable<Object> {
+    /**
+     * The value of the underlying property, replacing an empty provider by {@literal null}.
+     *
+     * This is required for allowing optional provider properties - all code which unpacks providers calls {@link Provider#get()} and would fail if an optional provider is passed.
+     * Returning {@literal null} from a {@link Callable} is ignored, and {@link PropertyValue} is a {@link Callable}.
+     */
     @Nullable
     @Override
     Object call();
 
+    /**
+     * The unprocessed value of the underlying property.
+     */
+    @Nullable
+    Object getValue();
```

Comment:
Could we use something like `Optional` to distinguish between the different kinds of values instead of introducing two separate methods?

Reply:
What do you mean by `something like `Optional`?

What we could do is that the consumers take care of the optional `Provider` handling for their use case. So `PropertyValue` would not implement `Callable` anymore.

That is doable, though a bigger effort. We can do this as a follow up - though not on `release` I think.

Reply:
Let's do that. In the meantime we could invest into better names for these two methods, too.

Reply:
Well, we can't do much about `call`, can we? `getValue` can be called `getUnprocessedValue`. WDYT?

Reply:
Let's go with that.

Reply:
Done.